### PR TITLE
Implement jumplist

### PR DESCRIPTION
**Pull request by glacambre**
This implements a jumplist similar to the one that can be found in vimperator/pentadactyl (bound to `<C-i>`/`<C-o>`).
The main difference is that reaching the end or the beginning of the jumplist won't load the previous/next page. This behavior seems more intuitive to me and is easier to implement.

I think we should bind jumpback and jumpnext to `<C-i>` and `<C-o>` once modifiers are supported.

### CODE REVIEW

**Comment by bovine3dom** [comment]
@glacambre where did we get to with this?

I think simply
```
if at start of jumplist and backwards: go back
else if at end of jumplist and forwards: go forward
otherwise: current code
```
would be fine. I'm happy to implement it if you're busy.

**Comment by bovine3dom** [comment]
Yeah, it would erase the jump list. I'm not sure I care a lot. There's no real rush - I just wanted to make sure that we didn't forget about the PR.

File: src/excmds.ts [hunk 1]
```diff
@@ -370,6 +370,58 @@ export function loggingsetlevel(logModule: string, level: string) {
...
+    jumpprev(-n)
+}
+
+/** Similar to Pentadactyl or vim's jump list.
+    Should be bound to <C-o> when modifiers are implemented
+*/
+//#content
+export function jumpprev(n = 1) {
+    JUMPCURRENT -= n
+    if (JUMPCURRENT < 0) return back(-JUMPCURRENT)
```
Comment (bovine3dom) [inline src/excmds.ts]: If there's no page to go back to, `JUMPCURRENT` gets negative, and you have to jumpnext more times than you'd expect to start going forwards again.

Forwards has a similar problem.

I'd suggest something like this:

```ts
export function jumpprev(n = 1) {
    if (JUMPCURRENT-n < 0) {
        return back(-JUMPCURRENT+n)
    }
    else if (JUMPCURRENT-n >= JUMPLIST.length){
        return forward(JUMPCURRENT - n - JUMPLIST.length + 1)
    }
    JUMPCURRENT -= n
    let p = JUMPLIST[JUMPCURRENT]
    JUMPED = true
    window.scrollTo(p.x, p.y)
}
```

File: src/excmds.ts [hunk 2]
```diff
@@ -370,6 +370,58 @@ export function loggingsetlevel(logModule: string, level: string) {
...
+    else if (JUMPCURRENT >= JUMPLIST.length) return forward(JUMPCURRENT - JUMPLIST.length + 1)
+    let p = JUMPLIST[JUMPCURRENT]
```
Comment (bovine3dom) [inline src/excmds.ts]: If you want to quickly move through history with this bound to `<C-{i,o}>`, we get all the way to here as `JUMPLIST` has zero length, and then `p` often ends up undefined as the page hasn't been open for 1000ms yet and we get an error popping up in the command line.

File: src/excmds.ts [hunk 3]
```diff
@@ -370,6 +370,58 @@ export function loggingsetlevel(logModule: string, level: string) {
...
+    clearTimeout(JUMPTIMEOUT)
+    JUMPTIMEOUT = setTimeout(() => {
+        JUMPLIST.push({ x: scrollEvent.pageX, y: scrollEvent.pageY })
+        JUMPCURRENT = JUMPLIST.length - 1
+    }, config.get("jumpdelay"))
+}
+
+//#content_helper
+document.addEventListener("scroll", addJump)
+addJump(new UIEvent("scroll"))
```
Comment (bovine3dom) [inline src/excmds.ts]: This needs `//#content_helper` above it (it should give you a build error without it...).

**Comment by glacambre** [comment]
I believe this PR is now ready to be merged. I gave up on implementing automatic tab storage (you can see my implementation here: https://github.com/glacambre/tridactyl/commit/2e7aac9307e99f306692caa507cfe47743102a79 , the main problem with it is terrible to use values from objects that are recursively proxied, you need to add "await" literally every time you access the a variable).

There's still what I'd consider a small bug when it comes to jumping to a location within a page by following a link (e.g. clicking on "further reading" on this page:  https://en.wiktionary.org/wiki/Test ): a jump to the very beginning of the page will be inserted in the jumplist. I don't know how to fix that and it doesn't break anything so I believe it shouldn't prevent this PR from being merged.

**Comment by bovine3dom** [comment]
>it is terrible to use values from objects that are recursively proxied, you need to add "await" literally every time you access the a variable

What's so bad about that?

TypeScript will tell you any time you try to use a Promise<T> instead of a T if you play your cards right.

**Comment by glacambre** [comment]
Well for example this:
```ts
let jumps = await tabstorage.get("jumps")
jumps.current -= 1
jumps[page].list.push({x: 0, y:0})
```
becomes this:
```ts
let jumps = await tabstorage.get("jumps")
(await jumps).current -= 1
(await (await (await jumps)[page]).list).push({x: 0, y:0})
```

**Comment by glacambre** [comment]
I realize that my message wasn't explicit enough. I gave up on implementing automatic tab storage but I still implemented manual storage (i.e. using browser.session.setTabValue/getTabValue) so the jumplist isn't lost when going to the previous/next page in history.

**Pull request merged**